flutter/lib/src/screenshot/screenshot_attachment_processor.dart

```diff
@@ -0,0 +1,77 @@
+import 'dart:async';
+import 'dart:typed_data';
+import 'dart:ui' as ui show ImageByteFormat;
+
+import 'package:flutter/rendering.dart';
+import 'package:sentry/sentry.dart';
+import 'package:sentry/sentry_private.dart';
+import '../integrations/native_app_start_integration.dart';
+import '../renderer/renderer.dart';
+import '../sentry_flutter_options.dart';
+import 'sentry_screenshot_widget.dart';
+
+// ignore: invalid_use_of_internal_member
+class ScreenshotAttachmentProcessor implements SentryClientAttachmentProcessor {
+  final SchedulerBindingProvider _schedulerBindingProvider;
+  final SentryFlutterOptions _options;
+
+  ScreenshotAttachmentProcessor(this._schedulerBindingProvider, this._options);
+
+  /// This is true when the SentryWidget is in the view hierarchy
+  bool get _attachScreenshot =>
+      sentryScreenshotWidgetGlobalKey.currentContext != null;
+
+  @override
+  Future<List<SentryAttachment>> processAttachments(
+      List<SentryAttachment> attachments, SentryEvent event) async {
+    if (event.exceptions == null &&
+        event.throwable == null &&
+        _attachScreenshot) {
+      return attachments;
+    }
+    final renderer = _options.rendererWrapper.getRenderer();
+    if (renderer != FlutterRenderer.skia &&
+        renderer != FlutterRenderer.canvasKit) {
+      return attachments;
+    }
+
+    final schedulerBinding = _schedulerBindingProvider();
+    if (schedulerBinding != null) {
+      final completer = Completer<Uint8List?>();
+      schedulerBinding.addPostFrameCallback((timeStamp) async {
+        final screenshot = await _createScreenshot();
+        completer.complete(screenshot);
+      });
+      final bytes = await completer.future;
```

Comment:
Could it be that this frame callback is never executed or executed way later again (seconds) and the code gets stuck in this place for a while?

Reply:
We fail in the `sentry_flutter_test.dart` tests, where we only check the renderer and install the screenshot integration. It only fails on chrome with a timeout.

Reply:
Not sure I understand your answer, let me try to rephrase my question.
We rely on that `addPostFrameCallback` always get executed otherwise `await completer.future` will wait forever.
Is there a possibility that the callback never gets executed or it takes a lot of time?
My question here is, what if there's no frame being rendered now or not in the next 5s? is the code stuck?
Maybe @ueman can help here.

Reply:
those should have only been run on vm, like the other tests.

Reply:
we are talking about different things, there's no testing here, its the processor, please read it thru.

Reply:
as far as i undestand the post frame callback should always be called exactly once. Any more insights here @ueman ?
We could also introduce a safeguard in the form of a timeout wehre we complete with an error and return.

Reply:
Ok, it might not be called when going to background according to this reddit user, haven't found this documented elsewhere yet.

https://www.reddit.com/r/FlutterDev/comments/rrt3b7/comment/hql0kcl/?utm_source=share&utm_medium=web2x&context=3

Reply:
The reason for using `addPostFrameCallback()` is that you have to wait until a frame is rendered to be able to take a screenshot. If you try to take a screenshot during build or rasterization, it will fail.
The callback is called after a frame is rendered, when the app is not in the foreground, there are of course no frames rendered at all.

That being said, in tests you may need to call `tester.binding.scheduleFrame()` in order to schedule a new frame. It kinda depends on how the test is set up.

Reply:
That sounds like the way to go, I'd avoid the `addPostFrameCallback`.
We try to take the screenshot but if it fails, it's fine, it's a best-effort feature.
I don't think we should block people's code for the sake of the screenshot.

Reply:
What could an approach without the callback look like? I checked out a [third party dependency](https://github.com/SachinGanesh/screenshot/blob/df14c5ccd52fcb5ccb568a0043419c9850438257/lib/screenshot.dart#L72), and they introduce a delay when calling the image API.

It seems that there is a limitation where you can't take a screenshot if `boundary.debugNeedsPaint` is true. The solution is to recursively call itself, which also seems kinda risky to me, as it potentially could lead to an endless loop. https://github.com/flutter/flutter/issues/22308

With this in mind, we could write our own small object which would handle these cases as a best effort.

If we use the `addPostFrameCallback` with a timeout, it should be short enough so that most devices have enough time to create the screenshot and not too long, as to not block our pipeline if we run into this issue repeatedly.

An approach with the timeout and recursive call should also be limited to a fixed number of retries where we bail out.

Reply:
I believe Manoel suggests to just use `image.toByteData()` and bail out if it fails.

The delay actually doesn't solve the problem, but just postpones it to a couple of moments later.

I also played around with removing the post frame callback callback back then
https://github.com/getsentry/sentry-dart/pull/576#discussion_r699349644 
Maybe it's worth double checking?

Reply:
<img width="719" alt="Bildschirmfoto 2022-11-08 um 15 37 56" src="https://user-images.githubusercontent.com/3984453/200593297-46fcc74f-4610-413f-8f42-c75697b18197.png">

It does produce a screenshot in the simulator, but the question is if this is flaky on devices.

Reply:
<img width="719" alt="Bildschirmfoto 2022-11-08 um 15 37 56" src="https://user-images.githubusercontent.com/3984453/200593297-46fcc74f-4610-413f-8f42-c75697b18197.png">

It does produce a screenshot in the simulator, but the question is if this is flaky on devices.

Reply:
the side effect of using `addPostFrameCallback` is much bigger than the feature itself.
I'd go with using `image.toByteData()` directly, wrapping in a try/catch and bail out if it's not possible, this feature is a best-effort.
if it works most of the time on release mode, we are good, lets's test using different types of errors, such as layout errors (flutter error), unhandled errors (zoned guard), etc.